### PR TITLE
Deprecate Misconfiguration Exception

### LINKED ISSUE #10918: [RFC] Revisit Usage of MisconfigurationException
## Proposed refactor
Replace Misconfiguration Exception with more diverse error types.

### Motivation

Currently we are using the `MisconfigurationException` for everything even if they are not directly related to the user's configuration. This means that the type of exception has no additional information about the kind of error. Python has some builtin exception types for that like `RuntimeError`, `ValueError`, `TypeError` etc. To gain more insights from the actual exception type, I propose to not use MisconfigurationException any longer and replace it with the other error types.

### Pitch
Have Lightning raising appropriate error types instead of just `MisconfigurationException` everywhere.

### Additional context

If we want to have some Lightning-Unique Exceptions we could still do something like 
```python
class LightningValueError(ValueError):
    pass

```
for the necessary error types and then raise these. This has the advantage that people could still have custom error handling also depending on types and statements like `except ValueError` would still catch those but they could also decide to only catch Lightning Errors.
______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

PS: Not sure under which part of the code to put it, since it probably touches all of them :D

cc @borda @justusschock @awaelchli @akihironitta @rohitgr7 @ananthsub @tchaton @carmocca 

EDIT: current usage of the `MisconfigurationException`: https://grep.app/search?q=MisconfigurationException&filter[repo][0]=PyTorchLightning/pytorch-lightning

### PR DESCRIPTION
**Pull request by calebbuffa**
## What does this PR do?
Replaces uses of `MisconfigurationException` with `_*Error` that inherets from a build-in `Exception`.

Fixes #10918

### Does your PR introduce any breaking changes? If yes, please list them.

No

## Before submitting

- [ X] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ X] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ X] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [X ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by calebbuffa** [comment]
This draft is just for code under `pytorch_lightning`. There are a few files under `lightning_lite` I haven't gotten into yet. I also haven't updated the tests, but I will of course do that. This draft is just to get feedback on the implementation and changes I've made. If they look good, I'll continue onto `lightning_lite` and the tests.

**Review by justusschock — COMMENTED** [review]
exception usage and metaclass look good to me. I just don't see a definition of the newly added Error types

**Comment by calebbuffa** [comment]
@justusschock thanks for reviewing! the error types can be found here: src/pytorch_lightning/utilities/exceptions.py

File: setup.py [hunk 1]
```diff
@@ -53,7 +53,7 @@
 
 from setuptools import setup
 
-_PACKAGE_NAME = os.environ.get("PACKAGE_NAME", "")
+_PACKAGE_NAME = ""
```
Comment (carmocca) [inline setup.py:56]: This was an accidental change, should be undone

File: examples/pl_loops/yielding_training_step.py [hunk 2]
```diff
@@ -64,7 +64,7 @@ def connect(self, **kwargs):
     def on_run_start(self, optimizers, kwargs):
         super().on_run_start(optimizers, kwargs)
         if not inspect.isgeneratorfunction(self.trainer.lightning_module.training_step):
-            raise MisconfigurationException("The `LightningModule` does not yield anything in the `training_step`.")
+            raise _RuntimeError("The `LightningModule` does not yield anything in the `training_step`.")
```
Comment (carmocca) [inline examples/pl_loops/yielding_training_step.py:67]: This is an exampple so you can just use `RuntimeError`

File: src/lightning_lite/utilities/exceptions.py [hunk 3]
```diff
@@ -12,6 +12,32 @@
...
+            raise TypeError(f"`{subclass.__name__}` should not be used for checks.")
+        if cls is MisconfigurationException:
+            rank_zero_deprecation(
+                f"`{MisconfigurationException.__name__}` is deprecated. Please check with `{real.__name__}` instead.",
+                stacklevel=stacklevel,
+            )
+        return cls in mro
+
+
+class MisconfigurationException(Exception, metaclass=_DeprecatedException):
```
Comment (carmocca) [inline src/lightning_lite/utilities/exceptions.py:38]: For files under `src/lightning_lite`, you can use the regular Python exceptions. No need to care about backwards compatibility.

cc @awaelchli 

The code you added here, should just be in `src/pytorch_lightning/utilities/exceptions.py`

Comment (calebbuffa) [reply src/lightning_lite/utilities/exceptions.py:38]: I'll revert this files' changes and move them under `pytorch_lightning`

File: examples/pl_loops/yielding_training_step.py (hunk 2 above)
Comment (calebbuffa) [reply examples/pl_loops/yielding_training_step.py:67]: :thumbsup:

**Comment by Borda** [comment]
@carmocca @awaelchli how is it going here? :chipmunk:

**Comment by carmocca** [comment]
Closing as this became stale. We are allowing breaking changes before the 2.0 release. So this PR could be continued without worrying about the backwards compatible deprecation classes. That would make it easier to land

Also it can be done in pieces, not everything needs to be updated at once.

Thank you for your time!
